pilot/model/controller_view.go

```diff
@@ -0,0 +1,580 @@
+	smallestSetLen := math.MaxInt32
+	setIdx := 0
+	for l, v := range labels {
+		valueKeysetMap := vl[l]
+		if len(valueKeysetMap) == 0 {
+			// Nothing matched at least one label name
+			return viewKeySet{}
+		}
+		valKeySet := valueKeysetMap[v]
+		lenKeySet := len(valKeySet)
+		if lenKeySet == 0 {
+			// There were no service keys for this label value
+			return viewKeySet{}
+		}
+		matchingSets[setIdx].keySet = valKeySet
+		if lenKeySet < smallestSetLen {
+			smallestSetLen = lenKeySet
+			if setIdx > 0 {
+				swappedMatchingSet := matchingSets[0]
+				matchingSets[0] = matchingSets[setIdx]
+				matchingSets[setIdx] = swappedMatchingSet
+			}
+		}
+		setIdx++
+	}
+	if countLabels == 1 {
+		return matchingSets[0].keySet
+	}
+	out := newViewKeySet(matchingSets[0].keySet)
+	for k := range out {
+		for setIdx := 1; setIdx < countLabels; setIdx++ {
+			_, found := matchingSets[setIdx].keySet[k]
+			if !found {
+				delete(out, k)
+				break
+			}
+		}
+	}
+	return out
+}
+
+// Given a label name, fetches a set of keys that have the label,
+// irrespective of what the label value may be.
+func (vl viewLabels) getAllKeysMatchingLabel(l string) viewKeySet {
+	valueKeysetMap := vl[l]
+	if len(valueKeysetMap) == 0 {
+		// Nothing matched at least one label name
+		return viewKeySet{}
+	}
+	out := viewKeySet{}
+	for _, keySet := range valueKeysetMap {
+		out.appendAll(keySet)
+	}
+	return out
+}
+
+// addLabel creates the reverse lookup by label name and label value for the key k.
+func (vl viewLabels) addLabel(k, labelName, labelValue string) {
+	valueKeysetMap, labelNameFound := vl[labelName]
+	if !labelNameFound {
+		valueKeysetMap = make(viewValues)
+		vl[labelName] = valueKeysetMap
+	}
+	keySet, labelValueFound := valueKeysetMap[labelValue]
+	if !labelValueFound {
+		keySet = make(viewKeySet)
+		valueKeysetMap[labelValue] = keySet
+	}
+	keySet[k] = true
+}
+
+// deleteLabel removes the key k from the reverse lookup of the label name and value
+func (vl viewLabels) deleteLabel(k string, labelName, labelValue string) {
+	valueKeysetMap, labelNameFound := vl[labelName]
+	if !labelNameFound {
+		return
+	}
+	keySet, labelValueFound := valueKeysetMap[labelValue]
+	if !labelValueFound {
+		return
+	}
+	delete(keySet, k)
+	if len(keySet) > 0 {
+		return
+	}
+	delete(valueKeysetMap, labelValue)
+	if len(valueKeysetMap) > 0 {
+		return
+	}
+	delete(vl, labelName)
+}
+
+// newKeySet creates a copy of fromKs that can be modified without altering fromKs
+func newViewKeySet(fromKs viewKeySet) viewKeySet {
+	out := make(viewKeySet, len(fromKs))
+	for k, v := range fromKs {
+		if v {
+			out[k] = v
+		}
+	}
+	return out
+}
+
+// appendAll appends all keys from fromKs to ks
+func (ks viewKeySet) appendAll(fromKs viewKeySet) {
+	for k, v := range fromKs {
+		if v {
+			ks[k] = v
+		}
+	}
+}
+
+// getNormalizedIP returns a normalized IPv6 addresses encoded in hex
+// given an IPv4 or IPv6 address
+func getNormalizedIP(address string) string {
```

Comment:
I don't think this makes sense - both IP and port have a well defined 'canonical' representation.
Ports are represented as decimal, IPs in their normal format. Why obfuscate them ?